docs/painless/painless-operators-boolean.asciidoc

```diff
@@ -0,0 +1,516 @@
+def dl = l0;                    // declares the def variable dl and sets it to l0
+boolean result;                 // declares the boolean variable result
+
+result = b0 !== b1;             // compares b0 to b1 and has a boolean result of true
+result = i !== f;               // compares i to f where i is promoted to float and has a boolean result of false
+result = b0 !== i;              // ERROR: a comparison between a boolean and a primitive numeric type is illegal
+result = i !== l0;              // ERROR: a comparison between a primitive numeric type and a reference type is illegal
+
+l0.add(1);                      // adds a constant int 1 to the List l0
+l1.add(1);                      // adds a constant int 1 to the ArrayList l1
+result = l0 !== l1;             // compares l0 to l1 and has a boolean result of true
+l0.add(1);                      // adds a constant int 1 to the List l0
+result = l0 !== l1;             // compares l0 to l1 and has a boolean result of true
+result = l1 !== l2;             // compares l1 to l2 and has a boolean result of false
+
+result = di0 !== di1;           // compares di0 to di1 and has a boolean result of true
+result = di0 !== i;             // compares di0 to i where i is promoted to def and has a boolean result of false
+
+result = dl !== l0;             // compares dl to l0 with a boolean result of false
+
+result = null !== dl;           // compares null to dl with a boolean result of true
+result = l1 !== null;           // compares null to l1 with a boolean result of true
+----
+
+[[boolean-xor-operator]]
+==== Boolean Xor
+
+Boolean xor will xor together two boolean expressions.  The table below shows what the resultant boolean value will be based on the two boolean expressions.
+
+|====
+||true|false
+|true|false|true
+|false|true|false
+|====
+
+The format starts with an expression, follows with the carrot operator, and finishes with an expression.
+
+*Grammar:*
+[source,ANTLR4]
+----
+boolean_xor: expression '^' expression;
+----
+
+Note that def types will be assumed to be of the boolean type.  Any def type evaluated at run-time that does not represent a boolean will result in an error.  Non-boolean expressions will result in an error.
+
+*Examples:*
+[source,Java]
+----
+boolean x = false;    // declares the boolean variable x and sets the constant boolean false
+boolean y = x ^ true; // declares the boolean variable y and sets it the result of x xor true
+def z = y ^ x;        // declares the def variable z and sets it to the result of y xor x
+----
+
+[[boolean-and-operator]]
+==== Boolean And
+
+Boolean and will and together two boolean expressions.  If the first expression is found to be false then it is known that the result will also be false, so evaluation of the second expression will be skipped.  The table below shows what the resultant boolean value will be based on the two boolean expressions.
+
```

Comment:
I'd probably go with, "Boolean `and` ANDs two boolean expressions." Is it worth adding, "and evaluates to true if both expressions are true"?

(Also corresponding changes for the following sections.)

Reply:
Changed, used caps.